api/v1alpha1/ingressnodefirewall_types.go

```diff
@@ -18,129 +18,139 @@ package v1alpha1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
 // IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
 type IngressNodeFirewallICMPRule struct {
-	// ICMPType define ICMP Type Numbers (RFC 792).
-	// +kubebuilder:validation:Required
+	// imcpType define ICMP Type Numbers (RFC 792).
+	// if its set the type field in the ICMP /ICMP6 header will match against this configured value otherwise its ignored.
+	// +kubebuilder:validation:Optional
 	// +kubebuilder:validation:Maximum:=255
 	// +kubebuilder:validation:Minimum:=0
-	// +optional
 	ICMPType uint8 `json:"icmpType,omitempty"`
 
-	// ICMPCode define ICMP Code ID (RFC 792).
-	// +kubebuilder:validation:Required
-	// +kubebuilder:validation:Maximum:=16
+	// icmpCode define ICMP Code ID (RFC 792).
+	// if its set the code field in the ICMP /ICMP6 header will match against this configured value otherwise its ignored.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum:=255
 	// +kubebuilder:validation:Minimum:=0
-	// +optional
 	ICMPCode uint8 `json:"icmpCode,omitempty"`
 }
 
 // IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
 type IngressNodeFirewallProtoRule struct {
-	//Ports can be a single port or range of ports using start-end range format
+	// ports defines either a single port or a range of ports to apply a protocol rule too.
+	// To filter a single port, set a single port as an integer value. For example ports: 80.
+	// To filter a range of ports, use a "start-end" range, string format. For example ports: "80-100".
 	//+kubebuilder:validation:Optional
-	// +optional
-	Ports string `json:"ports,omitempty"`
+	Ports intstr.IntOrString `json:"ports,omitempty"`
 }
 
-// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
-type IngressNodeFirewallProtocolRule struct {
-	// Order define order of execution of ingress firewall rules .
+// IngressNodeProtocolConfig is a discriminated union of protocol's specific configuration.
+// +union
+type IngressNodeProtocolConfig struct {
+	// protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +unionDiscriminator
 	// +kubebuilder:validation:Required
-	// +kubebuilder:validation:Minimum:=1
-	Order uint32 `json:"order"`
+	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
```

Comment:
When using a union, you have to have fields that match the discriminants. So protocolRule and icmpRule aren't valid

You need one for each of `icmp`, `icmpv6`, `tcp`, `udp` and `sctp` (note the fields I'm referring to are the json tags, the go types match the discriminant exactly)